Mathlib/GroupTheory/Index.lean

```diff
@@ -614,3 +614,30 @@ theorem index_center_le_pow [Finite (commutatorSet G)] [Group.FG G] :
 end FiniteIndex
 
 end Subgroup
+
+namespace MonoidHom
+
+open Finset
+
+variable {G H F : Type*} [Group G] [Fintype G] [Group H] [DecidableEq H] [MonoidHomClass F G H]
+
+@[to_additive]
+theorem card_fiber_of_mem_range (f : F) {x : H} (hx : x ∈ Set.range f) :
```

Comment:
You don't need the codomain to be a group. I have this code somewhere in the history of leanprover-community/mathlib#17653 (lost to a buggy merge):
```lean
@[to_additive] lemma monoid_hom.card_fiber_eq_of_mem_range {M} [monoid M] [decidable_eq M]
  (f : G →* M) {x y : M} (hx : x ∈ set.range f) (hy : y ∈ set.range f) :
  (univ.filter $ λ g, f g = x).card = (univ.filter $ λ g, f g = y).card :=
begin
  rcases ⟨hx, hy⟩ with ⟨⟨x, rfl⟩, y, rfl⟩,
  rcases mul_left_surjective x y with ⟨y, rfl⟩,
  conv_lhs { rw [← map_univ_equiv (equiv.mul_right y⁻¹), filter_map, card_map] },
  congr' 2 with g,
  simp only [(∘), equiv.to_embedding_apply, equiv.coe_mul_right, map_mul],
  rw [← f.coe_to_hom_units y⁻¹, map_inv, units.mul_inv_eq_iff_eq_mul, f.coe_to_hom_units]
```
Not sure if it depends on other modifications I made in that branch.

Reply:
UPD: have a look at [this](https://github.com/leanprover-community/mathlib/pull/17653/files#diff-7f1be6b9022b8228482801f42e5005df11252d1a068c3c1c8da0ee85078cc538R2014) section of that PR (`preimage`/`image` lemmas are better fomulated using pointwise smul).

Reply:
I just ported the above proof to Lean 4. It didn't depend on anything else in that old Lean 3 PR.